src/client/auth.js

```diff
@@ -1,60 +1,146 @@
+          options.connect(client)
+        }
+      }
+
+      if (!options.session && options.profilesFolder) {
+        let auths = await getLauncherProfiles()
+
+        let lowerUsername = options.username.toLowerCase()
+        let profile = Object.keys(auths.authenticationDatabase).find(key => 
+          auths.authenticationDatabase[key].username.toLowerCase() == lowerUsername
+          || Object.values(auths.authenticationDatabase[key].profiles)[0].displayName.toLowerCase() == lowerUsername
+        )
+
+        if (profile) {
+          let newUsername = auths.authenticationDatabase[profile].username
+          let uuid = Object.keys(auths.authenticationDatabase[profile].profiles)[0]
+          let displayName = auths.authenticationDatabase[profile].profiles[uuid].displayName
+          let newProfile = {
+            name: displayName,
+            id: uuid
+          };
+
+          options.session = {
+            accessToken: auths.authenticationDatabase[profile].accessToken,
+            clientToken: auths.clientToken,
+            selectedProfile: newProfile,
+            availableProfiles: [newProfile]
           }
-        })
+          options.username = newUsername
+        }
+      }
+
+      if (options.session) {
+        if (!skipValidation) {
+          yggdrasilClient.validate(options.session.accessToken, function (err) {
+            if (!err) { cb(null, options.session) } else {
+              yggdrasilClient.refresh(options.session.accessToken, options.session.clientToken, function (err, accessToken, data) {
+                if (!err) {
+                  cb(null, data)
+                } else if (options.username && options.password) {
+                  yggdrasilClient.auth({
+                    user: options.username,
+                    pass: options.password,
+                    token: clientToken,
+                    requestUser: true
+                  }, cb)
+                } else {
+                  cb(err, data)
+                }
+              })
+            }
+          })
+        } else {
+          // trust that the provided session is a working one
+          cb(null, options.session)
+        }
       } else {
-        // trust that the provided session is a working one
-        cb(null, options.session)
+        yggdrasilClient.auth({
+          user: options.username,
+          pass: options.password,
+          token: clientToken
+        }, cb)
       }
+
     } else {
-      yggdrasilClient.auth({
-        user: options.username,
-        pass: options.password,
-        token: clientToken
-      }, cb)
+      // assume the server is in offline mode and just go for it.
+      client.username = options.username
+      options.connect(client)
     }
-  } else {
-    // assume the server is in offline mode and just go for it.
-    client.username = options.username
-    options.connect(client)
-  }
+  })().then((res) => {}, (err) => {
+    console.warn("Authentication failed: \n", err)
```

Comment:
Printing error to the console is not a good approach. For instance it would break layout of mineflayer-dashboard.

Reply:
What should be done instead?